src/include/miopen/tripletmarginloss/problem_description.hpp

```diff
@@ -0,0 +1,199 @@
+    TensorDescriptor oDesc;
+
+    NetworkConfig MakeForwardNetworkConfig() const;
+};
+
+struct BackwardProblemDescription : ProblemDescriptionBase
+{
+
+    BackwardProblemDescription(const TensorDescriptor& aDesc_,
+                               const TensorDescriptor& pDesc_,
+                               const TensorDescriptor& nDesc_,
+                               const TensorDescriptor& dODesc_,
+                               const TensorDescriptor& dADesc_,
+                               const TensorDescriptor& dPDesc_,
+                               const TensorDescriptor& dNDesc_)
+        : aDesc(aDesc_),
+          pDesc(pDesc_),
+          nDesc(nDesc_),
+          dODesc(dODesc_),
+          dADesc(dADesc_),
+          dPDesc(dPDesc_),
+          dNDesc(dNDesc_)
+    {
+    }
+
+    const TensorDescriptor& GetADesc() const { return aDesc; }
+    const TensorDescriptor& GetPDesc() const { return pDesc; }
+    const TensorDescriptor& GetNDesc() const { return nDesc; }
+    const TensorDescriptor& GetdODesc() const { return dODesc; }
+    const TensorDescriptor& GetdADesc() const { return dADesc; }
+    const TensorDescriptor& GetdPDesc() const { return dPDesc; }
+    const TensorDescriptor& GetdNDesc() const { return dNDesc; }
+
+    bool IsSameType() const
+    {
+        if(!checkSameType(aDesc, pDesc) || !checkSameType(aDesc, nDesc))
+            MIOPEN_THROW(
+                miopenStatusBadParm,
+                "Triplet Margin Loss: Anchor, Positive, Negative tensor must have same type.");
+        if(!checkSameType(aDesc, dADesc) || !checkSameType(pDesc, dPDesc) ||
+           !checkSameType(nDesc, dNDesc))
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "Triplet Margin Loss: Gradient tensors (excluding Output gradient) must "
+                         "share a same type with Anchor, Positive and Negative tensor.");
+        return true;
+    }
+
+    bool IsRightLength() const
+    {
+        if(aDesc.GetSize() != 2 || pDesc.GetSize() != 2 || nDesc.GetSize() != 2)
+            MIOPEN_THROW(
+                miopenStatusBadParm,
+                "Triplet Margin Loss: Anchor, Positive, Negative tensor must have 2 dimensions.");
+        if(dADesc.GetSize() != 2 || dPDesc.GetSize() != 2 || dNDesc.GetSize() != 2)
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "Triplet Margin Loss: Gradient tensors (excluding Output gradient) must "
+                         "have 2 dimensions.");
+        if(dODesc.GetSize() != 1)
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "Triplet Margin Loss: Output gradient tensor must have 1 dimension.");
+        if(!checkSameLength(aDesc, pDesc) || !checkSameLength(aDesc, nDesc))
+            MIOPEN_THROW(
+                miopenStatusBadParm,
+                "Triplet Margin Loss: Anchor, Positive, Negative tensor sizes do not match.");
+        return true;
+    }
+
+    bool IsReduced() const
```

Comment:
Do you need both IsReduced and IsUnreduced? Aren't they exclusive?

Reply:
I believe that I need both of them, because there aren't any parts in my code validate the element size or dim[0] of output and grad output tensors. For example, an output tensor and an anchor tensor with shape (2) and (10,10) may fail both `IsReduced` and `IsUnreduced` but they can pass all other checks.

Reply:
Okay. You're using them in IsApplicable. But if there's !IsReduced() && !IsUnreduced() case, doesn't it need to be considered invalid input?
That kinds of checks should be done at the constructor of `ProblemDescription`.
Check this review comment. https://github.com/ROCm/MIOpen/pull/2623#discussion_r1450730047

And if you do so, I think you can use one of IsReduced and IsUnreduced and remove other one.

Reply:
I have moved those checks which contain `MIOPEN_THROW` to problem description constructor. It includes those in `IsSameType()` and `IsRightLength()` which made me removed those 2 functions along with `IsUnreduced()`.